mock/mock.go

```diff
@@ -798,6 +828,23 @@ func (args Arguments) Diff(objects []interface{}) (string, int) {
 				differences++
 				output = fmt.Sprintf("%s\t%d: FAIL:  type %s != type %s - %s\n", output, i, reflect.TypeOf(t).Name(), reflect.TypeOf(actual).Name(), actualFmt)
 			}
+		} else if reflect.TypeOf(expected) == reflect.TypeOf((*FunctionalOptionsArgument)(nil)) {
+			name := expected.(*FunctionalOptionsArgument).name
+			t := expected.(*FunctionalOptionsArgument).value
+			tName := reflect.TypeOf(t).Name()
+			if name != reflect.TypeOf(actual).String() {
```

Comment:
We should be derive the `name` field on `FunctionalOptionsArgument` with a bit more reflection and simplify the `FunctionalOptions()` signature

```suggestion
			t := expected.(*FunctionalOptionsArgument).value

			var name string
			tValue := reflect.ValueOf(t)
			if tValue.Len() > 0 {
				name = "[]" + reflect.TypeOf(tValue.Index(0).Interface()).String()
			}

			tName := reflect.TypeOf(t).Name()
			if name != reflect.TypeOf(actual).String() && tValue.Len() != 0 {
```